### ISSUE TITLE
Exporting schemas from package breaks `instanceof`

### ISSUE THREAD

**Issue by ahmafi** [colinhacks/zod#2241]
When I export a schema from a package and use that schema in another package, the `instanceof` operator doesn't work on `ZodError` or `Schema`.

How I'm using the exported schemas:

```ts
import { mySchema } from 'mylib';
import { z } from 'zod';

console.log(mySchema instanceof z.Schema); // false

try {
  mySchema.parse('');
} catch (error) {
  console.log(error instanceof z.ZodError); // false
}

// -----------------------------

const localSchema = z.object({});

console.log(localSchema instanceof z.Schema); // true

try {
  localSchema.parse('');
} catch (error) {
  console.log(error instanceof z.ZodError); // true
}
```

I have also created a [demo project](https://github.com/ahmafi/zoderror-showcase). 

I call the package that I create schemas in `mylib` and the package that I'm using those, is called `myprogram`. I'm using `npm link` to add `mylib` to `myprogram` package. And also, I'm transpiling the `mylib` with `npx tsc`. Then you can test the code in `myprogram` with `npx tsx index.ts`.

**Comment by joepetrillo** [comment]
I am also having this same issue!

**Comment by ahmafi** [comment]
Thanks, it solved the issue. However, if I set the `"type": "module"` in the `myprogram`'s `package.json` then the issue occurs again.

**Comment by gtflip** [comment]
I ran into this problem with one of our internal libraries.  It's because it checks the prototype and the instances from different copies of the library have different prototypes.  It's the reason we have `Array.isArray()`, which works across different "realms", like different iframes.

I solved it by overriding the `[Symbol.hasinstance]` function.  I'm still not sure it was the best way to do it, but it solved the problem and didn't require changing all the times I used `instanceof` in other libraries.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof

**Comment by KhaledElmorsy** [comment]
Same problem here. Tried setting it as a peer dependency in the consumed package and made sure that the workspace only has one `/zod` folder, and still cant solve this.

**Comment by rodeyseijkens** [comment]
So yea because of version differences between the packages.

Not sure if there is a good way to handle this besides having the same version.
Maybe it would be a solution to have the following?

```ts
import { mySchema } from 'mylib';
import { z } from 'zod';

try {
  mySchema.parse('');
} catch (error) {
  console.log(error instanceof mySchema.ZodError); // true
}
```

**Comment by muratgozel** [comment]
Temporarily did `if (error.constructor.name === 'ZodError'/* error instanceof z.ZodError */)`

**Comment by mmkal** [comment]
This can also happen even with a single zod installation, if one part of the code accesses the commonjs export (`main` in package.json) and another part accesses the esm export (`module`).

Issue in trpc-cli: https://github.com/mmkal/trpc-cli/issues/7

If anyone's interested, a non-bullet-proof-but-in-most-cases-good-enough fix is to look at the constructor name as above, but you should also account for sub-classes:

```ts
export const looksLikeInstanceof = <T>(value: unknown, target: new (...args: any[]) => T): value is T => {
  let current = value?.constructor
  do {
    if (current?.name === target.name) return true
    current = Object.getPrototypeOf(current) as Function
  } while (current?.name)
  return false
}
```

PR in trpc-cli: https://github.com/mmkal/trpc-cli/pull/8

CC @colinhacks in case you know of some reason why the method above wouldn't be a good idea for some reason. Given zod ships both commonjs and esm, so has the [dual-package hazard](https://nodejs.org/api/packages.html#dual-package-hazard), it would be nice if it used `Symbol.hasInstance` as @gtflip suggests to avoid this out of the box.

Also related: https://github.com/evanw/esbuild/issues/3333

**Comment by colinhacks** [comment]
Oh wow. Embarrassingly I wasn't aware of `Symbol.hasInstance` and it is such a game changer. I've seen multiple issues relating to dual package issues and was looking into some way to avoid ESM/CJS dual package (with an ESM wrapper that re-exports ESM, as recommended in the Node.js docs). 

But of course there are always scenarios where multiple versions of Zod will be floating around. 

In Zod 4 all schemas will have a `kind` property (e.g. `kind = Symbol.for("ZodString")`) and I'll implement `hasInstance` with a simple symbol comparison.

**Comment by gtflip** [comment]
Yeah, it took me a bit of searching for a solution before I found it.  I also ran into it with one of our internal packages, which is why I had to find a solution.  😃 

I had difficulty finding example implementations for my use case, but I store an object at a key in each instance of my class and in `Symbol.hasInstance` I check to see if that key exists, the key contains an object of the right shape, and the values in the shape are what I expect.  If it is, I check that the return of `this.getType()` is what I expect it to be.

**Comment by andresgutgon** [comment]
Related. Hard lesson learned here 😂 
https://github.com/IdoPesok/zsa/issues/121#issuecomment-2171686993

**Comment by tianhuil** [comment]
Hi @colinhacks: we're running up against this issue.  Would you accept a contribution to `Symbol.hasInstance`?

**Comment by MarijnMensinga** [comment]
Note to self (and hopefully others):  
What fixed the problems for us was to import zod consistently like this: `const {z} = require('zod');` instead of mixing imports with `const z = require('zod');`

**Comment by bompi88** [comment]
I'm not sure if this works for all use cases, but I ended up with adding a util and running this in the local "service" before checking `instanceof`. The catch is that I need to know that it is a schema beforehand, but that is usually the case anyways.

```ts
export const wrapWithLocalZodInstance = <T extends z.ZodTypeAny>(schema: T): T => {
  let wrappedSchema;
  if ('shape' in schema && typeof schema.shape === 'object') {
    wrappedSchema = z.object({ ...schema.shape });
  } else {
    wrappedSchema = z.unknown().and(schema);
  }
  Object.setPrototypeOf(schema, Object.getPrototypeOf(wrappedSchema));
  return schema;
};

const schema = z.object({ test: z.string() });
const wrappedSchema = wrapWithLocalZodInstance(schema);

schema instanceof z.ZodType // false
wrappedSchema instanceof z.ZodType // true
```

**Comment by agazibaric** [comment]
@ahmafi  

To address the issue, could you try the following steps?  

Instead of bundling the code inside your library and exporting the bundled version (`dist/index.js`), update your library to export the `index.ts` file directly. For example, your `package.json` could look like this:  

```json
{
  "name": "mylib",
  "main": "index.ts",
  "dependencies": {
    "zod": "^3.21.4"
  },
  "devDependencies": {
    "typescript": "^5.0.2"
  }
}
```  

With this change, `instanceof` operator should work properly inside of `myprogram` package.

**Comment by kpozin** [comment]
Note that this also breaks some logic within Zod itself. For example, `.required(mask)` [relies on][1] `instanceof ZodOptional` to find the inner type of an optional field.

```ts
        while (newField instanceof ZodOptional) {
          newField = (newField as ZodOptional<any>)._def.innerType;
        }
```

If the field type was imported from another package and made optional, the `while` loop short-circuits and the field is never set to _required_.

[1]: https://github.com/colinhacks/zod/blob/e2b9a5f9ac67d13ada61cd8e4b1385eb850c7592/src/types.ts#L3003-L3005

**Comment by CodeSmith32** [comment]
If I read through this correctly, this is not an issue with Zod; it's just a Duplicate Dependency issue, which can happen with any package. Furthermore, Zod v4 *even still* fixes this issue (🎉) because it uses string-labeled trait-based instanceof checks by overriding `[Symbol.hasInstance]` as mentioned.

**I posit this issue should be closed.**
